File: chazm-model/src/main/java/runtimemodels/chazm/model/id/IdFactory.java
package runtimemodels.chazm.model.id;

import runtimemodels.chazm.api.id.UniqueId;

/**
 * The {@linkplain IdFactory} interface defines the APIs for constructing {@linkplain UniqueId}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public interface IdFactory {

    /**
     * Constructs an {@linkplain UniqueId}.
     *
     * @param <T> the type that this {@linkplain UniqueId} will be associated.
     * @param <U> the type of the identifier for this {@linkplain UniqueId}.
     * @param clazz the type of the {@linkplain UniqueId}.
     * @param id    the {@linkplain Class} id for the {@linkplain UniqueId}.
     * @return an {@linkplain UniqueId}.
     */
    <T, U> UniqueId<T> build(Class<T> clazz, Class<U> id);

    /**
     * Constructs an {@linkplain UniqueId}.
     *
     * @param <T> the type that this {@linkplain UniqueId} will be associated.
     * @param clazz the type of the {@linkplain UniqueId}.
     * @param id    the <code>long</code> id for the {@linkplain UniqueId}.
     * @return an {@linkplain UniqueId}.
     */
    <T> UniqueId<T> build(Class<T> clazz, Long id);

    /**
     * Constructs an {@linkplain UniqueId}.
     *
     * @param <T> the type that this {@linkplain UniqueId} will be associated.
     * @param clazz the type of the {@linkplain UniqueId}.
     * @param id    the {@linkplain String} id for the {@linkplain UniqueId}.
     * @return an {@linkplain UniqueId}.
     */
    <T> UniqueId<T> build(Class<T> clazz, String id);

}

File: chazm-model/src/main/java/runtimemodels/chazm/model/id/IdFactory.kt
package runtimemodels.chazm.model.id

import runtimemodels.chazm.api.id.UniqueId

/**
 * The [IdFactory] interface defines the APIs for constructing [UniqueId]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
interface IdFactory {

    /**
     * Constructs an [UniqueId].
     *
     * @param <T> the type that this [UniqueId] will be associated.
     * @param <U> the type of the identifier for this [UniqueId].
     * @param clazz the type of the [UniqueId].
     * @param id    the [Class] id for the [UniqueId].
     * @return an [UniqueId].
    </U></T> */
    fun <T, U> build(clazz: Class<T>, id: Class<U>): UniqueId<T>

    /**
     * Constructs an [UniqueId].
     *
     * @param <T> the type that this [UniqueId] will be associated.
     * @param clazz the type of the [UniqueId].
     * @param id    the `long` id for the [UniqueId].
     * @return an [UniqueId].
    </T> */
    fun <T> build(clazz: Class<T>, id: Long?): UniqueId<T>

    /**
     * Constructs an [UniqueId].
     *
     * @param <T> the type that this [UniqueId] will be associated.
     * @param clazz the type of the [UniqueId].
     * @param id    the [String] id for the [UniqueId].
     * @return an [UniqueId].
    </T> */
    fun <T> build(clazz: Class<T>, id: String): UniqueId<T>

}
